src/node/internal/internal_http_outgoing.ts

```diff
@@ -111,14 +115,24 @@ class MessageBuffer {
     encoding: WrittenDataBufferEntry['encoding'],
     callback: WrittenDataBufferEntry['callback']
   ): boolean {
-    this.#data.push({
+    const entry: WrittenDataBufferEntry = {
       data,
       length: data?.length ?? 0,
       encoding,
       callback,
-      written: false,
-    });
-    this._flush();
+      written: true,
+    };
+
+    if (this.#corked === 0) {
+      this.#onWrite(this.#index++, entry);
+      callback?.();
```

Comment:
Even in this case the callback is expected to be run async.

```suggestion
      queueMicrotask(() => callback?.());
```

For example, in Node.js:
```js
const { Writable } = require('stream');
const w = new Writable({ write(d, e, cb) { console.log('.'); cb(); }});
w.write('hello', () => console.log('!'));
console.log('written');
```

Outputs:

```
.
written
!
```